### PR TITLE
demux_playlist: Use alphanum sorting for files

### PR DESCRIPTION
**Pull request by ksv1986**
Using algorithm by David Koelle
http://www.davekoelle.com/alphanum.html

I agree that my changes can be relicensed to LGPL 2.1 or later.

### CODE REVIEW

File: demux/demux_playlist.c [hunk 1]
```diff
@@ -270,9 +271,73 @@ static bool scan_dir(struct pl_parser *p, char *path,
     return true;
 }
 
+/* http://www.davekoelle.com/alphanum.html */
```
Comment (wm4) [inline demux/demux_playlist.c]: License violation.

File: demux/demux_playlist.c [hunk 2]
```diff
@@ -270,9 +271,73 @@ static bool scan_dir(struct pl_parser *p, char *path,
...
+        NUMBER,
+    } mode = STRING;
+
+    while (*l && *r) {
+        if (mode == STRING) {
+            char l_char, r_char;
+            while ((l_char = *l) && (r_char = *r)) {
+                // check if this are digit characters
+                const int l_digit = isdigit(l_char);
+                const int r_digit = isdigit(r_char);
```
Comment (wm4) [inline demux/demux_playlist.c]: This is a crashing bug.

**Comment by sandsmark** [comment]
not sure if the posix wcscoll is usable on all platforms mpv supports, if so just using that is probably the easiest.

otherwise here's the collation implementations in qt for various platforms. might want to look at them, most of them just use plain native C APIs on the platforms, and the qt code is lgpl 2.1, so no licensing issues:

https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qcollator_icu.cpp?h=v5.6.0
https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qcollator_win.cpp?h=v5.6.0
https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qcollator_posix.cpp?h=v5.6.0
https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qcollator_macx.cpp?h=v5.6.0

**Comment by rossy** [comment]
If it turns out that strcoll/wcscoll in Linux and macOS supports natural sorting and strcoll/wcscoll in Windows doesn't, we could always add a wrapper in osdep/io.c, like we do with most of the other "POSIX" functions that Windows' C library exposes.

**Comment by sandsmark** [comment]
seems like the qt implementation enables natural sorting of numbers manually with the native icu/macos/windows backends (sets various flags), so a simple wrapper of CompareString{,Ex} with the SORT_DIGITSASNUMBERS flag is probably the easiest for windows at least.

**Comment by sandsmark** [comment]
Looking around more, this implementation actually looks pretty nice (result-wise, the code isn't very pretty): https://github.com/GNOME/glib/blob/master/glib/gunicollate.c#L517-L561

File: demux/demux_playlist.c (hunk 1 above)
Comment (kevmitch) [reply demux/demux_playlist.c]: Isn't this MIT licensed?

Comment (kevmitch) [reply demux/demux_playlist.c]: The answer is that the actual copyright notice must be preserved as per MIT.

**Comment by wm4** [comment]
Closing as this went nowhere.

**Pull request closed without merging**